### DISCUSSION TITLE
Aptly 1.6.0 Release Preparation Update!

### DISCUSSION THREAD

**Discussion by neolynx** [aptly-dev/aptly#1345] [Announcements]
Dear Aptly community, 

The time has come for a new Aptly release, and we’d like to update you on recent developments and what you can expect from the upcoming version.

### 💼 What’s Been Happening on the Maintainers' Side

Over the past few months, maintainers have worked diligently to improve the development and CI environments while reviving long-standing pull requests and issues. 

Key internal achievements include:

- **Reliable System Tests**
    - Fix, improve and enable more tests
    - Use of S3 mirrors for reliability
- **Updated Pipeline**
    - Use go 1.22
    - Build with go or native debian tools (replace goxc)
    - Improved Debian support (see below)
- **Updated Development Environment**
    - Use docker containers
    - Improve local unit and system testing

These updates allow us to focus more efficiently on the remaining issues and PRs while also enhancing the developer experience and lowering the barriers to start contributing.

### 📝 Changes Included in the upcoming Release

To name a few, the following improvements have been merged and will be in the next release:

- Support updating the components of a published repo
- Support publishing multiple distributions (-multi-dist)
- Support Etcd Database 
- Allow slash (/) in distribution names
- Support for storing the "local" pool on Azure 
- Improved or added new REST APIs
- Fix publish concurrency and improve performance
- Improved Mirroring
- fix download throttling
- fix resuming package downloads
- fix ignoring signatures
- Fix packages dependency resolution (Virtual Packages, version numbers in Provides)
- Improved S3 support and performance
- Fix race condition with goleveldb
- …

### 📦 New Debian Packages

The pipeline is now using the debian toolchain to build debian packages (replacing goxz) in order to build for more distributions and architectures. Also binaries for MacOs, FreeBSD and Linux are now built with go directly and will be provided for the release version.

Aptly CI (aka nightly) builds now available as debian packages for the following distributions and architectures:

- Debian 10 (buster), (11) bullseye, (12) bookworm, testing/trixie
- Ubuntu 20.04 (focal), 22.04 (jammy), 24.04 (noble)
- Architectures: amd64, i386, arm64, armhf

The following additional Debian packages are available now:

- **aptly-api** 	(REST API systemd service)
    - based on aptly-api in debian
    - uses global /etc/aptly.conf config file
    - Note: Automatically migrates /etc/aptly-api.conf to /etc/aptly.conf during upgrades from the official Debian package.
- **aptly-dbg** 
    - debug symbols

Currently only CI builds are available, but the upcoming release will deliver the same packages

This also changed the APT repositories structure, please make sure to update as described below.

### 🗃️ New APT repository

To try the latest development version, use the following APT source to install Aptly CI builds:
```
deb [signed-by=/etc/apt/keyrings/aptly.asc] http://repo.aptly.info/ci DIST main
```
where DIST may be one of: `buster`, `bullseye`, `bookworm`, `trixie` (testing), `focal`, `jammy`, `noble`

This replaces the current `http://repo.aptly.info/nightly-DIST` apt repositories. The official releases in future will be using `http://repo.aptly.info/release` as APT URL.

**Repo Key**

The repo key did not change, if can be obtained as follows if needed:
```
sudo wget http://repo.aptly.info/pubkey.txt -O /etc/apt/keyrings/aptly.asc
```

### 🙏 Special Thanks!

We want to extend a huge thank you to all the new maintainers who have stepped up and contributed their valuable time and expertise. Your efforts have been instrumental in making this upcoming release possible! 

To all the contributors and users who have been patiently waiting for their Issues and PRs to be addressed – we hear you, and we appreciate your understanding. We’re working diligently to catch up, and we hope this release will reflect our commitment to improving the project. 

Special thanks also to everyone who has contributed by responding to issues, resolving them, or helping improve the documentation and website.

Further work is still needed in these areas, and any assistance would be greatly appreciated. If you're able, you can also support Aptly through [GitHub Sponsorships](https://github.com/sponsors/aptly-dev) to help cover maintenance and hosting costs.

### 📅 Timeline

With just a few more PRs to merge and issues to resolve, Aptly will soon be ready for its next release.

We’re aiming for the next release within 2–3 months. During this time, feel free to test the latest CI builds, contribute, or discuss new features.

Thank you all for your continued support and contributions! 🚀

Best,
The Maintainers Team

**Comment by ju2wheels** [comment]
Is there a reason why the libc6 dependencies for the 20.04 packages are different for AMD64 vs ARM64. I know what im trying to do is "technically" not in the supported list by installing the 20.04 package on 18.04 but for AMD64 this seems to at least install fine but I have to go through the testing still to see if theres other issues. But for 20.04 ARM64 package the dependency is way different... for example the 20.04 AMD64 package dep for `libc6 is (>= 2.3.4)` while for ARM64 it is `libc6:arm64 (>= 2.28)`. Why is there such a large difference in dependency version on the same distro?

**Reply by neolynx** [reply]
Indeed:
```
$ dpkg -I aptly_1.6.0~rc1+20241202144730.3fad1965_amd64.deb | grep libc
 Depends: libc6 (>= 2.3.4), bzip2, xz-utils, gpgv, gpg
$ dpkg -I aptly_1.6.0~rc1+20241202144730.3fad1965_arm64.deb | grep libc
 Depends: libc6:arm64 (>= 2.31), bzip2, xz-utils, gpgv, gpg
```
I see a different version for arm though.. are you using the CI APT  sources above?

I think the go linker determines the libc, as we are cross compiling for arm64, maybe this is the reason ? Could you try building native arm64 build to compare ?

Running on 18.04 should be ok, as long as libc matches, so I understand your worry.

**Reply by ju2wheels** [reply]
Ill double checked that I didnt accidentally grab the wrong version. I was able to get the 20.04 AMD64 package to work fine on 18.04 but not the ARM64. Ill grab it again and repack it with `ar` to modify the control file deps and see if it works (if not ill try the same with the buster package which should be closer to 18.04 than the 20.04 package).

Im not sure about how to rebuild, is that process described in the contrib doc?

**Reply by neolynx** [reply]
the contrib doc is being updated, see https://github.com/aptly-dev/aptly/blob/improve/documentation/CONTRIBUTING.md#local-development-setup

bulding on arm64 (bookworm/backports):
```
$ dpkg -I build/aptly_1.6.0~rc1+20241202144730.3fad1965_arm64.deb | grep libc
 Depends: libc6 (>= 2.34), bzip2, xz-utils, gpgv, gpg
```
